xapian-letor/api/letor_internal.cc

```diff
@@ -271,23 +231,149 @@ Letor::Internal::prepare_training_file(const string & queryfile, const string &
+	if (querystr.empty()) {
+	    throw LetorParseError("Empty query string in query file at line:" + Xapian::Internal::str(query_count));
+	}
+
+	string qq = querystr;
+	istringstream iss(querystr);
+	string title = "title:";
+	while (iss) {
+	    string t;
+	    iss >> t;
+	    if (t.empty())
+		break;
+	    string temp;
+	    temp.append(title);
+	    temp.append(t);
+	    temp.append(" ");
+	    temp.append(qq);
+	    qq = temp;
+	}
+	Xapian::Query query = parser.parse_query(qq,
+						 parser.FLAG_DEFAULT|
+						 parser.FLAG_SPELLING_CORRECTION);
+	Xapian::Enquire enquire(letor_db);
+	enquire.set_query(query);
+	Xapian::MSet mset = enquire.get_mset(0, msetsize);
 
-    if(!myfile1.good()){
-	cout << "No Query file found"<<endl;
-	throw FileNotFound();
+	std::vector<FeatureVector> fvv_mset = flist.create_feature_vectors(mset, query, letor_db);
+	std::vector<FeatureVector> fvv_qrel;
+	int k = 0;
+	for (Xapian::MSetIterator i = mset.begin(); i != mset.end(); ++i) {
+	    Xapian::Document doc = i.get_document();
+	    int label = getlabel(doc, qid);
+	    if (label != -1) { // only add FeatureVector which is found in the qrel file
+		fvv_mset[k].set_label(label);
+		fvv_qrel.push_back(fvv_mset[k]);
+	    }
+	    ++k;
+	}
+	write_to_file(fvv_qrel, qid, train_file);
     }
+    myfile1.close();
+    train_file.close();
+}
 
+void
+Letor::Internal::letor_score(const std::string & query_file, const std::string & qrel_file,
+			     const std::string & model_file, const std::string & output_file,
+			     Xapian::doccount msetsize, const Xapian::FeatureList & flist)
+{
+    // Load ranker model
+    ranker->load_model_from_file(model_file.c_str());
 
-    while (!myfile1.eof()) {           // reading all the queries line by line from the query file
+    Xapian::SimpleStopper mystopper(sw, sw + sizeof(sw) / sizeof(sw[0]));
+    Xapian::Stem stemmer("english");
+    Xapian::QueryParser parser;
+    parser.add_prefix("title", "S");
+    parser.add_prefix("subject", "S");
+    parser.set_database(letor_db);
+    parser.set_default_op(Xapian::Query::OP_OR);
+    parser.set_stemmer(stemmer);
+    parser.set_stemming_strategy(Xapian::QueryParser::STEM_SOME);
+    parser.set_stopper(&mystopper);
 
-	getline(myfile1, str1);
+    this->qrel = load_relevance(qrel_file);
+
+    string str1;
+    ifstream queryfile;
+    queryfile.open(query_file.c_str(), ios::in);
+    if(!queryfile.good()) {
+	throw Xapian::FileNotFoundError("No Query file found. Check path.");
+    }
 
+    ofstream out_file;
+    out_file.open(output_file);
+
+    double total_score = 0;
+    int num_queries = 0;
+    while (!queryfile.eof()) {
+	getline(queryfile, str1);
 	if (str1.empty()) {
 	    break;
 	}
+	++num_queries;
 
-	string qid = str1.substr(0, (int)str1.find(" "));
-	string querystr = str1.substr((int)str1.find("'")+1, (str1.length() - ((int)str1.find("'") + 2)));
+	char * str;
+	char * x1;
+	x1 = const_cast<char*>(str1.c_str());
+	str = strtok(x1, " ");
```

Comment:
This call to `strtok()` will modify the string pointed to by `x1`, which I'm fairly sure is undefined behaviour.

It's usually better to use methods of `std::string` to manipulate a `std::string` rather than the string functions C++ inherits from C, but that's especially true of `strtok()` which modifies the string it works on and isn't even reentrant (`strtok_r()` is reentrant, but is specified by POSIX rather than the ISO C standard, so is less portable).

You can split a string into words at spaces using methods of `std::string` like this:

```
        size_t j = 0;
        while (j < s.size()) {
            size_t i = s.find_first_not_of(' ', j);
            if (i == string::npos) break;
            j = s.find_first_of(' ', i);
            token.push_back(s.substr(i, j - i));
        }
```

(The second argument of `substr()` can be larger than the length of the string, which happens if there's a word at the end of the string, since then `j` is `string::npos`).

You can also specify several characters instead of a single one, so to split at space or tab:

```
s.find_first_not_of(" \t", j);
```

etc.